hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/processors/StreamToStreamJoinP.java

```diff
@@ -325,22 +331,44 @@ public boolean saveToSnapshot() {
                 }
             }
 
-            Stream<Entry<?, ?>> leftBufferStream = buffer[0].content()
+            leftBufferStream = buffer[0].content()
                     .stream()
                     .map(row -> entry(
-                            ObjectArrayKey.project(row, joinInfo.leftEquiJoinIndices()),
+                            joinInfo.isEquiJoin() ? project(row, joinInfo.leftEquiJoinIndices()) : row.hashCode(),
                             new BufferSnapshotValue(row, unusedEventsTracker.contains(row), 0)
                     ));
 
-            Stream<Entry<?, ?>> rightBufferStream = buffer[1].content()
+            rightBufferStream = buffer[1].content()
                     .stream()
                     .map(row -> entry(
-                            ObjectArrayKey.project(row, joinInfo.rightEquiJoinIndices()),
+                            joinInfo.isEquiJoin() ? project(row, joinInfo.rightEquiJoinIndices()) : row.hashCode(),
                             new BufferSnapshotValue(row, unusedEventsTracker.contains(row), 1)
                     ));
 
+            if (!joinInfo.isEquiJoin()) {
+                // Note : details are described in TDD : docs/design/sql/15-stream-to-stream-join.md
+                if (joinInfo.isRightOuter() && processorIndex == 0) {
+                    leftBufferStream = buffer[0].content()
```

Comment:
conditional overwriting of `leftBufferStream` and `rightBufferStream` is harder to read. It should be easy to initialize them to correct value according to `joinInfo.isEquiJoin()`